### PR TITLE
Add Prow job to detect CVEs in ACK supported AWS service controllers

### PR DESCRIPTION
**Pull request by michaelhtm**
Description of changes:

Currently ACK does not have an automated way of detecting issues in our controllers.
This prow job will allow us to detect any vulnerabilities in our controller images 
and notify us early enough so we can take action on time.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: ack-build-tools
Comment (a-hilaly) [inline ack-build-tools:1]: I think you forgot a binary here :) do you want to add `./ack-build-tools` to the gitignore?

File: prow/jobs/images/Dockerfile.scan-controllers-cve [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+# Start a new stage from scratch
+FROM alpine:latest
+
+RUN apk add --no-cache wget tar
+
+RUN export TRIVY_VERSION=0.54.1 && \
+    wget https://github.com/aquasecurity/trivy/releases/download/v${TRIVY_VERSION}/trivy_${TRIVY_VERSION}_Linux-64bit.tar.gz -O - | tar -xz && \
+    chmod +x trivy && \
+    mv trivy /usr/local/bin/
```
Comment (a-hilaly) [inline prow/jobs/images/Dockerfile.scan-controllers-cve:21]: nit: Eventually we want to make TRIVY_VERSION an arg as well.

File: prow/jobs/tools/cmd/command/scan_controllers_cve.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	scanCveCMD.PersistentFlags().StringVar(
+		&OptGithubIssueRepo, "github-issues-repo", "community", "repository where the github issue will be created",
+	)
+	rootCmd.AddCommand(scanCveCMD)
+}
+
+func scanControllersCve(cmd *cobra.Command, args []string) error {
+
+	log.Printf("Retreiving AWS services list from %s\n", OptJobsConfigPath)
+	services, err := getAckServiceList(OptJobsConfigPath)
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve.go:52]: nit: s/getAckServiceList/getACKServices/

File: prow/jobs/tools/cmd/command/scan_controllers_cve.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		return err
+	}
+
+	log.Println("Scanning All controllers for CVEs")
+	vulnerabilitiesControllers, uniqueVulnerabilities, err := getControllersCveSummaries(controllerTagsMap)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("We currently have %d unique CVEs", len(vulnerabilitiesControllers))
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve.go:69]:
```suggestion
	log.Printf("We found %d unique CVEs across all the controllers\n", len(vulnerabilitiesControllers))
```

File: prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+	"strings"
+	"time"
+
+	"gopkg.in/yaml.v3"
+
+	"github.com/aws-controllers-k8s/test-infra/prow/jobs/tools/cmd/command/generator"
+)
+
+// CHANGE when there's a new service controller
+const controllersCount = 43
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go:30]: Do we need to bump this value each time we release a new controller?

File: prow/jobs/tools/cmd/command/upgrade_go_version.go [hunk 5]
```diff
@@ -57,7 +57,7 @@ func runUpgradeGoVersion(cmd *cobra.Command, args []string) error {
 
 	log.Printf("Successfully listed go versions from %s\n", OptGoEcrRepository)
 
-	highestEcrGoVersion, err := findHighestGoVersion(ecrGoVersions)
+	highestEcrGoVersion, err := findHighestTagVersion(ecrGoVersions)
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/upgrade_go_version.go:60]: Good call!

File: prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+	start := time.Now()
+
+	for {
+		if time.Since(start) > timeout {
+			return nil, fmt.Errorf("timeout to get repository tags after %d seconds", timeout)
+		}
+		if retries > maxRetries {
+			return nil, fmt.Errorf("timeout to get repository tags after %d retries", retries)
+		}
+		time.Sleep(1 * time.Second)
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go:103]: Maybe sleep after the api call?

File: prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+		tags, err := listRepositoryTags(input)
+		if err != nil {
+			if strings.Contains(err.Error(), "429") {
+				retries++
+				continue
+			} else {
+				return nil, fmt.Errorf("continue")
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go:110]:
```suggestion
				return nil, err
```

File: prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+					FixedVersion:     vulnerability.FixedVersion,
+					Severity:         vulnerability.Severity,
+					Title:            vulnerability.Title,
+					Type:             result.Type,
+				}
+				controllers, ok := vulnIdToControllers[vulnerability.VulnerabilityId]
+				if ok {
+					vulnIdToControllers[vulnerability.VulnerabilityId] = append(controllers, controller)
+				} else {
+					vulnIdToControllers[vulnerability.VulnerabilityId] = make([]string, 0, 50)
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go:183]: why 50?

File: prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+	return controllerTagsMap, nil
+}
+
+func getControllersCveSummaries(controllerTagsMap map[string]string) (map[string][]string, map[string]CVESummary, error) {
+
+	app := "trivy"
+
+	cveSummaries := make(map[string]CVESummary)
+	vulnIdToControllers := make(map[string][]string)
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go:145]:
```suggestion
	detectedVulnerabilities := make(map[string][]string)
```

File: prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go (hunk 8 above)
Comment (a-hilaly) [reply prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go:183]:
```suggestion
					vulnIdToControllers[vulnerability.VulnerabilityId] = []string{controller}
```

File: prow/jobs/tools/cmd/command/upgrade_go_version_helper.go [hunk 10]
```diff
@@ -42,12 +42,12 @@ func listRepositoryTags(repository string) ([]string, error) {
...
 	tags, err := client.ListRepositoryTags(repository)
 	if err != nil {
-		return nil, fmt.Errorf("cannot list repositories in %s. %s", OptGoEcrRepository, err)
+		return nil, fmt.Errorf("cannot list repositories in %s. %s", repository, err)
 	}
 	return tags, nil
 }
 
-func findHighestGoVersion(tags []string) (string, error) {
+func findHighestTagVersion(tags []string) (string, error) {
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/upgrade_go_version_helper.go:50]: Can we have some unit tests for this function?

File: prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+		if err != nil {
+			return nil, nil, fmt.Errorf("%s\n%s", stdout, err)
+		}
+
+		var trivyOutput *TrivyOutput
+		if err = json.Unmarshal(stdout, &trivyOutput); err != nil {
+			return nil, nil, err
+		}
+
+		for _, result := range trivyOutput.Results {
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go:168]: we could wrap all this logic into a function ... `getCVESummaries` ?

File: prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+		return nil, err
+	}
+
+	return config.AWSServices, nil
+}
+
+// Sending a request to ecrpublic too many times causes a 429 Too Many Requests error
+// ListRepositoryWithRetries reattempts requests for maxRetries or timeout duration
+// when such error is received
+func listRepositoryTagsWithRetries(input string, maxRetries int, timeout time.Duration) ([]string, error) {
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go:92]: FYI this is a nice read https://go.dev/doc/comment

File: prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go (hunk 11 above)
Comment (a-hilaly) [reply prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go:168]: Then we can have unit tests for it :)

File: prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+		args := []string{
+			"image",
+			"--format",
+			"json",
+			"-q",
+			fmt.Sprintf(scanControllerImageFormat, controller, tag),
+		}
+		cmd := exec.Command(app, args...)
+		stdout, err := cmd.CombinedOutput()
+		//TODO: return error for now, revist later
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go:158]: What's the plan for later?

File: prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+	}
+
+}
+
+func getControllersLatestTags(services []string) (map[string]string, error) {
+	controllerTagsMap := make(map[string]string)
+
+	for _, service := range services {
+		// Sleeping to avoid StatusCode: 429 Too Many Requests error
+		controllerRepoTags, err := listRepositoryTagsWithRetries(fmt.Sprintf(ecrPublicControllerImageFormat, service), 10, 15*time.Second)
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go:123]:
```suggestion
		controllerRepoTags, err := listRepositoryTagsWithRetries(fmt.Sprintf(ecrPublicControllerImageFormat, service), 10, 1*time.Minute)
```

**Review by a-hilaly — COMMENTED** [review]
Great work on this @michaelhtm !! I have a few questions and suggestions below

File: ack-build-tools
Comment (michaelhtm) [reply ack-build-tools:1]: Good catch

File: prow/jobs/tools/cmd/command/scan_controllers_cve_tests.go [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+	}
+}
+
+func TestGetCveSummaries(t *testing.T) {
+
+	assert := assert.New(t)
+
+	type args struct {
+		controller              string
+		results             []Result
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve_tests.go:79]: This file need a `gofmt` round

File: prow/jobs/tools/cmd/command/scan_controllers_cve_tests.go [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+						"gobinary",
+					},
+				},
+				cveSummaries:            map[string]CVESummary{},
+				detectedVulnerabilities: map[string][]string{},
+			},
+			wantCveSummariesLength: 1,
+			wantDetectedVulnerabilitiesLength: 1,
+		},
+	}
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve_tests.go:111]: I think what you probably want to also test is the aggregation of CVEs. Thinking a test with 3 controller, 2 common CVEs, and one that is only impacting one controller

File: prow/jobs/tools/cmd/command/scan_controllers_cve_tests.go [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+package command
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFindHighestTagVersion(t *testing.T) {
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve_tests.go:23]: Nice tests!

File: prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+		}
+		if retries > maxRetries {
+			return nil, fmt.Errorf("timeout to get repository tags after %d retries", retries)
+		}
+		tags, err := listRepositoryTags(repository)
+		time.Sleep(1 * time.Second)
+		if err != nil {
+			if strings.Contains(err.Error(), "429") {
+				retries++
+				continue
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/scan_controllers_cve_helper.go:110]: I meant you only want to sleep if 429 is returned

**Review by a-hilaly — APPROVED** [review]
Good stuff Michael! I have some other comments, please address them in a follow up PR.
/lgtm

**Pull request merged**